// @trendmicro/react-portal/lib/Portal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Portal = (_temp2 = _class = function (_PureComponent) {
    _inherits(Portal, _PureComponent);

    function Portal() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Portal);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Portal.__proto__ || Object.getPrototypeOf(Portal)).call.apply(_ref, [this].concat(args))), _this), _this.node = null, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Portal, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            if (!this.node) {
                this.node = document.createElement('div');
                this.node.setAttribute('data-reactportal', '');

                if (this.props.node) {
                    this.props.node.appendChild(this.node);
                } else {
                    document.body.appendChild(this.node);
                }
            }
            this.componentDidUpdate();
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            if (this.node) {
                _reactDom2.default.unmountComponentAtNode(this.node);
                if (this.node.parentNode) {
                    this.node.parentNode.removeChild(this.node);
                }
                this.node = null;
            }
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
            var props = _extends({}, this.props);
            delete props.node;

            _reactDom2.default.render(_react2.default.createElement('div', props), this.node);
        }
    }, {
        key: 'render',
        value: function render() {
            return null;
        }
    }]);

    return Portal;
}(_react.PureComponent), _class.propTypes = {
    node: _propTypes2.default.any
}, _temp2);
exports.default = Portal;
